{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "e7b93413",
   "metadata": {},
   "source": [
    "## Required Pip downloads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "be6abad3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[33mWARNING: Ignoring invalid distribution ~cipy (/Users/joshuageorgedai/anaconda3/lib/python3.11/site-packages)\u001b[0m\u001b[33m\n",
      "\u001b[0mRequirement already satisfied: btclib in /Users/joshuageorgedai/anaconda3/lib/python3.11/site-packages (2023.7.12)\n",
      "Requirement already satisfied: btclib-libsecp256k1 in /Users/joshuageorgedai/anaconda3/lib/python3.11/site-packages (from btclib) (0.4.0)\n",
      "Requirement already satisfied: cffi>=1.0.0 in /Users/joshuageorgedai/anaconda3/lib/python3.11/site-packages (from btclib-libsecp256k1->btclib) (1.15.1)\n",
      "Requirement already satisfied: pycparser in /Users/joshuageorgedai/anaconda3/lib/python3.11/site-packages (from cffi>=1.0.0->btclib-libsecp256k1->btclib) (2.21)\n",
      "\u001b[33mWARNING: Ignoring invalid distribution ~cipy (/Users/joshuageorgedai/anaconda3/lib/python3.11/site-packages)\u001b[0m\u001b[33m\n",
      "\u001b[0m\n",
      "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m A new release of pip is available: \u001b[0m\u001b[31;49m24.0\u001b[0m\u001b[39;49m -> \u001b[0m\u001b[32;49m25.1.1\u001b[0m\n",
      "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m To update, run: \u001b[0m\u001b[32;49mpip install --upgrade pip\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "# !pip install ecdsa\n",
    "# !pip install btclib\n",
    "# !pip install python-bitcoinlib\n",
    "# !pip install bitcoinlib"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bce3c6b4",
   "metadata": {},
   "source": [
    "### btclib test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "202094bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "from btclib.ecc.ssa import sign, verify\n",
    "from btclib.hashes import sha256\n",
    "from btclib.ecc import dsa\n",
    "\n",
    "# Generate a private/public key pair\n",
    "private_key, public_key = dsa.gen_keys()\n",
    "#gen_keys randomly generates the keys\n",
    "#public key is a tuple of large numbers\n",
    "#private key is a large number\n",
    "\n",
    "# Message signing\n",
    "message = \"Hello Bitcoin\"\n",
    "#message.encode() encodes message in bytes\n",
    "msg_hash = sha256(message.encode())\n",
    "#sha256 output is in bytes, matches other sha256 generators\n",
    "signature = sign(msg_hash, private_key)\n",
    "#signatures gives the r, s, and Elliptical curve value\n",
    "\n",
    "# Verify the signature\n",
    "assert verify(msg_hash, public_key, signature)\n",
    "print(\"Schnorr signature verified successfully!\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "34b1f157",
   "metadata": {},
   "source": [
    "## python-bitcoinlib tests\n",
    "#### NEED to run Bitcoin node for this to work"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "2d00c270",
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "Cookie file unusable ([Errno 2] No such file or directory: '/Users/joshuageorgedai/Library/Application Support/Bitcoin/.cookie') and rpcpassword not specified in the configuration file: '/Users/joshuageorgedai/Library/Application Support/Bitcoin/bitcoin.conf'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "File \u001b[0;32m~/anaconda3/lib/python3.11/site-packages/bitcoin/rpc.py:192\u001b[0m, in \u001b[0;36mBaseProxy.__init__\u001b[0;34m(self, service_url, service_port, btc_conf_file, timeout, connection)\u001b[0m\n\u001b[1;32m    191\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 192\u001b[0m     \u001b[38;5;28;01mwith\u001b[39;00m \u001b[38;5;28mopen\u001b[39m(cookie_file, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mr\u001b[39m\u001b[38;5;124m'\u001b[39m) \u001b[38;5;28;01mas\u001b[39;00m fd:\n\u001b[1;32m    193\u001b[0m         authpair \u001b[38;5;241m=\u001b[39m fd\u001b[38;5;241m.\u001b[39mread()\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: '/Users/joshuageorgedai/Library/Application Support/Bitcoin/.cookie'",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[3], line 4\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mbitcoin\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mrpc\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m RawProxy\n\u001b[1;32m      3\u001b[0m \u001b[38;5;66;03m# Connect to local bitcoin node via RPC\u001b[39;00m\n\u001b[0;32m----> 4\u001b[0m p \u001b[38;5;241m=\u001b[39m RawProxy()\n\u001b[1;32m      6\u001b[0m \u001b[38;5;66;03m# Make an RPC call\u001b[39;00m\n\u001b[1;32m      7\u001b[0m info \u001b[38;5;241m=\u001b[39m p\u001b[38;5;241m.\u001b[39mgetblockchaininfo()\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.11/site-packages/bitcoin/rpc.py:315\u001b[0m, in \u001b[0;36mRawProxy.__init__\u001b[0;34m(self, service_url, service_port, btc_conf_file, timeout, **kwargs)\u001b[0m\n\u001b[1;32m    309\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m__init__\u001b[39m(\u001b[38;5;28mself\u001b[39m,\n\u001b[1;32m    310\u001b[0m              service_url\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m,\n\u001b[1;32m    311\u001b[0m              service_port\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m,\n\u001b[1;32m    312\u001b[0m              btc_conf_file\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m,\n\u001b[1;32m    313\u001b[0m              timeout\u001b[38;5;241m=\u001b[39mDEFAULT_HTTP_TIMEOUT,\n\u001b[1;32m    314\u001b[0m              \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs):\n\u001b[0;32m--> 315\u001b[0m     \u001b[38;5;28msuper\u001b[39m(RawProxy, \u001b[38;5;28mself\u001b[39m)\u001b[38;5;241m.\u001b[39m\u001b[38;5;21m__init__\u001b[39m(service_url\u001b[38;5;241m=\u001b[39mservice_url,\n\u001b[1;32m    316\u001b[0m                                    service_port\u001b[38;5;241m=\u001b[39mservice_port,\n\u001b[1;32m    317\u001b[0m                                    btc_conf_file\u001b[38;5;241m=\u001b[39mbtc_conf_file,\n\u001b[1;32m    318\u001b[0m                                    timeout\u001b[38;5;241m=\u001b[39mtimeout,\n\u001b[1;32m    319\u001b[0m                                    \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.11/site-packages/bitcoin/rpc.py:199\u001b[0m, in \u001b[0;36mBaseProxy.__init__\u001b[0;34m(self, service_url, service_port, btc_conf_file, timeout, connection)\u001b[0m\n\u001b[1;32m    196\u001b[0m             authpair \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;132;01m%s\u001b[39;00m\u001b[38;5;124m:\u001b[39m\u001b[38;5;132;01m%s\u001b[39;00m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;241m%\u001b[39m (conf[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mrpcuser\u001b[39m\u001b[38;5;124m'\u001b[39m], conf[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mrpcpassword\u001b[39m\u001b[38;5;124m'\u001b[39m])\n\u001b[1;32m    198\u001b[0m         \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m--> 199\u001b[0m             \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mCookie file unusable (\u001b[39m\u001b[38;5;132;01m%s\u001b[39;00m\u001b[38;5;124m) and rpcpassword not specified in the configuration file: \u001b[39m\u001b[38;5;132;01m%r\u001b[39;00m\u001b[38;5;124m'\u001b[39m \u001b[38;5;241m%\u001b[39m (err, btc_conf_file))\n\u001b[1;32m    201\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    202\u001b[0m     url \u001b[38;5;241m=\u001b[39m urlparse\u001b[38;5;241m.\u001b[39murlparse(service_url)\n",
      "\u001b[0;31mValueError\u001b[0m: Cookie file unusable ([Errno 2] No such file or directory: '/Users/joshuageorgedai/Library/Application Support/Bitcoin/.cookie') and rpcpassword not specified in the configuration file: '/Users/joshuageorgedai/Library/Application Support/Bitcoin/bitcoin.conf'"
     ]
    }
   ],
   "source": [
    "from bitcoin.rpc import RawProxy\n",
    "\n",
    "# Connect to local bitcoin node via RPC\n",
    "p = RawProxy()\n",
    "\n",
    "# Make an RPC call\n",
    "info = p.getblockchaininfo()\n",
    "\n",
    "print(\"Blockchain info:\", info)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8a530468",
   "metadata": {},
   "source": [
    "### python-bitcoinlib tests"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "0c8f3658",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Initial Confirmed Balance: 399.99985370 BTC\n",
      "Successfully sent 1.0 BTC!\n",
      "Transaction ID: 7e18b2f979da23c1ab2282154608ce63fa492b1f510121056910edefa1e4aff4\n",
      "Transaction confirmed by mining one block.\n",
      "New wallet balance: 449.99983290 BTC\n"
     ]
    }
   ],
   "source": [
    "# FIRST, boot up terminal, run Bitcoin Core, and create a wallet called \"myaddress\"\n",
    "from bitcoin.rpc import RawProxy\n",
    "# Connect directly to your wallet \"my address\"\n",
    "rpc_url = 'http://joshuageorgedai:333777000@127.0.0.1:18443/wallet/myaddress'\n",
    "proxy = RawProxy(service_url=rpc_url)\n",
    "\n",
    "# Create a new address to send coins to (for testing)\n",
    "receiver_address = proxy.getnewaddress()\n",
    "\n",
    "# Amount to send (e.g., 1 BTC)\n",
    "amount_to_send = 1.0\n",
    "\n",
    "# Send the Bitcoin\n",
    "# Wallet balance does not immediately reflect changes because you need 100 blocks mined\n",
    "# To see the effects of mining. Also, proxy is essentially transfering bitcoin to itself.\n",
    "try:\n",
    "    balances = proxy.getbalances()\n",
    "    initial_balance = balances['mine']['trusted']\n",
    "    print(f\"Initial Confirmed Balance: {initial_balance} BTC\")\n",
    "\n",
    "    txid = proxy.sendtoaddress(receiver_address, amount_to_send)\n",
    "    print(f\"Successfully sent {amount_to_send} BTC!\")\n",
    "    print(f\"Transaction ID: {txid}\")\n",
    "\n",
    "    # Mine a block to confirm transaction (only necessary for regtest), gives wallet that mined this block 50 btc\n",
    "    proxy.generatetoaddress(1, receiver_address)\n",
    "    print(\"Transaction confirmed by mining one block.\")\n",
    "\n",
    "    # Confirm balance update\n",
    "    # Transaction occurred properly if small transaction fee is taken\n",
    "    updated_balance = proxy.getbalance()\n",
    "    print(f\"New wallet balance: {updated_balance} BTC\")\n",
    "\n",
    "except Exception as e:\n",
    "    print(\"An error occurred:\", e)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "84330726",
   "metadata": {},
   "source": [
    "## bitcoinlib tests"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "051e73ee",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Address: bc1q9gd9crjq7evgqlq6r0av8qlmgm3jmw2jvu4fry\n",
      "Public Key: <WalletKey(key_id=6, name=address index 0, wif=zpub6unJwCYvS9MJiZjq8v58fEfvmf6WmqXq46EP7qgqV6GvuhZHfTSq8Ps5DhW9SmYv3JDQM6uEmVfhFwQAZqf37FAf8K1JFYWEQBWCd2j49cq, path=m/84'/0'/0'/0/0)>\n",
      "Private Key (WIF): zpub6unJwCYvS9MJiZjq8v58fEfvmf6WmqXq46EP7qgqV6GvuhZHfTSq8Ps5DhW9SmYv3JDQM6uEmVfhFwQAZqf37FAf8K1JFYWEQBWCd2j49cq\n"
     ]
    }
   ],
   "source": [
    "from bitcoinlib.wallets import Wallet, wallet_create_or_open\n",
    "\n",
    "\n",
    "#TEST Wallet\n",
    "# Create/Open a new wallet (testnet)\n",
    "wallet = wallet_create_or_open('MyTestWallet', network='regtest')\n",
    "\n",
    "# Generate a new key/address\n",
    "# key = wallet.new_key()\n",
    "\n",
    "#Get current key\n",
    "key = wallet.get_key()\n",
    "print(\"Address:\", key.address)\n",
    "print(\"Public Key:\", key.public())\n",
    "print(\"Private Key (WIF):\", key.wif)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "1980834f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "No UTXOs found. Wallet has no available funds.\n"
     ]
    }
   ],
   "source": [
    "# Get UTXOs (unspent outputs)\n",
    "utxos = wallet.utxos()\n",
    "\n",
    "# Print details of each UTXO\n",
    "if utxos:\n",
    "    print(f\"UTXOs for wallet '{wallet.name}':\\n\")\n",
    "    for utxo in utxos:\n",
    "        print(f\"Transaction ID: {utxo.txid}\")\n",
    "        print(f\"Output index: {utxo.output_n}\")\n",
    "        print(f\"Address: {utxo.address}\")\n",
    "        print(f\"Value: {utxo.value} satoshis ({utxo.value / 1e8} BTC)\")\n",
    "        print(f\"Confirmations: {utxo.confirmations}\")\n",
    "        print('-' * 40)\n",
    "else:\n",
    "    print(\"No UTXOs found. Wallet has no available funds.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9ffe555a",
   "metadata": {},
   "outputs": [
    {
     "ename": "TransactionError",
     "evalue": "Network for output address tb1qlj64u6fqutr0xue85kl55fx0gt4m4urun25p7q is different from transaction network. bitcoin not in ['testnet', 'testnet4', 'signet']",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTransactionError\u001b[0m                          Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[17], line 10\u001b[0m\n\u001b[1;32m      7\u001b[0m receiver_address \u001b[38;5;241m=\u001b[39m other_key\u001b[38;5;241m.\u001b[39maddress\n\u001b[1;32m      9\u001b[0m \u001b[38;5;66;03m# Send amount (e.g., 0.0001 BTC)\u001b[39;00m\n\u001b[0;32m---> 10\u001b[0m tx \u001b[38;5;241m=\u001b[39m wallet\u001b[38;5;241m.\u001b[39msend_to(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mtb1qlj64u6fqutr0xue85kl55fx0gt4m4urun25p7q\u001b[39m\u001b[38;5;124m\"\u001b[39m, amount\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m0.0001 BTC\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[1;32m     12\u001b[0m \u001b[38;5;66;03m#Send testnet bitcoin back to address tb1qlj64u6fqutr0xue85kl55fx0gt4m4urun25p7q\u001b[39;00m\n\u001b[1;32m     14\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mTransaction sent! TXID:\u001b[39m\u001b[38;5;124m\"\u001b[39m, tx\u001b[38;5;241m.\u001b[39mtxid)\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.11/site-packages/bitcoinlib/wallets.py:4395\u001b[0m, in \u001b[0;36mWallet.send_to\u001b[0;34m(self, to_address, amount, input_key_id, account_id, network, fee, min_confirms, priv_keys, locktime, broadcast, number_of_change_outputs, random_output_order, replace_by_fee)\u001b[0m\n\u001b[1;32m   4352\u001b[0m \u001b[38;5;250m\u001b[39m\u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m   4353\u001b[0m \u001b[38;5;124;03mCreate transaction and send it with default Service objects :func:`services.sendrawtransaction` method.\u001b[39;00m\n\u001b[1;32m   4354\u001b[0m \n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   4391\u001b[0m \u001b[38;5;124;03m:return WalletTransaction:\u001b[39;00m\n\u001b[1;32m   4392\u001b[0m \u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m   4394\u001b[0m outputs \u001b[38;5;241m=\u001b[39m [(to_address, amount)]\n\u001b[0;32m-> 4395\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39msend(outputs, input_key_id\u001b[38;5;241m=\u001b[39minput_key_id, account_id\u001b[38;5;241m=\u001b[39maccount_id, network\u001b[38;5;241m=\u001b[39mnetwork, fee\u001b[38;5;241m=\u001b[39mfee,\n\u001b[1;32m   4396\u001b[0m                  min_confirms\u001b[38;5;241m=\u001b[39mmin_confirms, priv_keys\u001b[38;5;241m=\u001b[39mpriv_keys, locktime\u001b[38;5;241m=\u001b[39mlocktime, broadcast\u001b[38;5;241m=\u001b[39mbroadcast,\n\u001b[1;32m   4397\u001b[0m                  number_of_change_outputs\u001b[38;5;241m=\u001b[39mnumber_of_change_outputs, random_output_order\u001b[38;5;241m=\u001b[39mrandom_output_order,\n\u001b[1;32m   4398\u001b[0m                  replace_by_fee\u001b[38;5;241m=\u001b[39mreplace_by_fee)\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.11/site-packages/bitcoinlib/wallets.py:4323\u001b[0m, in \u001b[0;36mWallet.send\u001b[0;34m(self, output_arr, input_arr, input_key_id, account_id, network, fee, min_confirms, priv_keys, max_utxos, locktime, broadcast, number_of_change_outputs, random_output_order, replace_by_fee)\u001b[0m\n\u001b[1;32m   4319\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m input_arr \u001b[38;5;129;01mand\u001b[39;00m max_utxos \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(input_arr) \u001b[38;5;241m>\u001b[39m max_utxos:\n\u001b[1;32m   4320\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m WalletError(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mInput array contains \u001b[39m\u001b[38;5;132;01m%d\u001b[39;00m\u001b[38;5;124m UTXO\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124ms but max_utxos=\u001b[39m\u001b[38;5;132;01m%d\u001b[39;00m\u001b[38;5;124m parameter specified\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;241m%\u001b[39m\n\u001b[1;32m   4321\u001b[0m                       (\u001b[38;5;28mlen\u001b[39m(input_arr), max_utxos))\n\u001b[0;32m-> 4323\u001b[0m transaction \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mtransaction_create(output_arr, input_arr, input_key_id, account_id, network, fee,\n\u001b[1;32m   4324\u001b[0m                                       min_confirms, max_utxos, locktime, number_of_change_outputs,\n\u001b[1;32m   4325\u001b[0m                                       random_output_order, replace_by_fee)\n\u001b[1;32m   4326\u001b[0m transaction\u001b[38;5;241m.\u001b[39msign(priv_keys)\n\u001b[1;32m   4327\u001b[0m \u001b[38;5;66;03m# Calculate exact fees and update change output if necessary\u001b[39;00m\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.11/site-packages/bitcoinlib/wallets.py:3953\u001b[0m, in \u001b[0;36mWallet.transaction_create\u001b[0;34m(self, output_arr, input_arr, input_key_id, account_id, network, fee, min_confirms, max_utxos, locktime, number_of_change_outputs, random_output_order, replace_by_fee)\u001b[0m\n\u001b[1;32m   3951\u001b[0m         \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(addr, WalletKey):\n\u001b[1;32m   3952\u001b[0m             addr \u001b[38;5;241m=\u001b[39m addr\u001b[38;5;241m.\u001b[39mkey()\n\u001b[0;32m-> 3953\u001b[0m         transaction\u001b[38;5;241m.\u001b[39madd_output(value, addr, change\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m)\n\u001b[1;32m   3955\u001b[0m srv \u001b[38;5;241m=\u001b[39m Service(network\u001b[38;5;241m=\u001b[39mnetwork, wallet_name\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mname, providers\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mproviders, cache_uri\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mdb_cache_uri,\n\u001b[1;32m   3956\u001b[0m               strict\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mstrict)\n\u001b[1;32m   3958\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m locktime \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39manti_fee_sniping:\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.11/site-packages/bitcoinlib/transactions.py:1940\u001b[0m, in \u001b[0;36mTransaction.add_output\u001b[0;34m(self, value, address, public_hash, public_key, lock_script, spent, output_n, encoding, spending_txid, spending_index_n, strict, change)\u001b[0m\n\u001b[1;32m   1938\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m value \u001b[38;5;241m!=\u001b[39m \u001b[38;5;241m0\u001b[39m:\n\u001b[1;32m   1939\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m TransactionError(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mOutput value for OP_RETURN script must be 0\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[0;32m-> 1940\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39moutputs\u001b[38;5;241m.\u001b[39mappend(Output(value\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mint\u001b[39m(value), address\u001b[38;5;241m=\u001b[39maddress, public_hash\u001b[38;5;241m=\u001b[39mpublic_hash,\n\u001b[1;32m   1941\u001b[0m                            public_key\u001b[38;5;241m=\u001b[39mpublic_key, lock_script\u001b[38;5;241m=\u001b[39mlock_script, spent\u001b[38;5;241m=\u001b[39mspent, output_n\u001b[38;5;241m=\u001b[39moutput_n,\n\u001b[1;32m   1942\u001b[0m                            encoding\u001b[38;5;241m=\u001b[39mencoding, spending_txid\u001b[38;5;241m=\u001b[39mspending_txid, spending_index_n\u001b[38;5;241m=\u001b[39mspending_index_n,\n\u001b[1;32m   1943\u001b[0m                            strict\u001b[38;5;241m=\u001b[39mstrict, change\u001b[38;5;241m=\u001b[39mchange, network\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mnetwork\u001b[38;5;241m.\u001b[39mname))\n\u001b[1;32m   1944\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m output_n\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.11/site-packages/bitcoinlib/transactions.py:700\u001b[0m, in \u001b[0;36mOutput.__init__\u001b[0;34m(self, value, address, public_hash, public_key, lock_script, spent, output_n, script_type, witver, encoding, spending_txid, spending_index_n, strict, change, witness_type, network)\u001b[0m\n\u001b[1;32m    698\u001b[0m network_guesses \u001b[38;5;241m=\u001b[39m address_dict[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mnetworks\u001b[39m\u001b[38;5;124m'\u001b[39m]\n\u001b[1;32m    699\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mnetwork\u001b[38;5;241m.\u001b[39mname \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m network_guesses:\n\u001b[0;32m--> 700\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m TransactionError(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mNetwork for output address \u001b[39m\u001b[38;5;132;01m%s\u001b[39;00m\u001b[38;5;124m is different from transaction network. \u001b[39m\u001b[38;5;132;01m%s\u001b[39;00m\u001b[38;5;124m not \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    701\u001b[0m                            \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124min \u001b[39m\u001b[38;5;132;01m%s\u001b[39;00m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;241m%\u001b[39m (\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_address, \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mnetwork\u001b[38;5;241m.\u001b[39mname, network_guesses))\n\u001b[1;32m    702\u001b[0m \u001b[38;5;66;03m# if address_dict['network'] and self.network.name != address_dict['network']:\u001b[39;00m\n\u001b[1;32m    703\u001b[0m \u001b[38;5;66;03m#     raise TransactionError(\"Address %s is from %s network and transaction from %s network\" %\u001b[39;00m\n\u001b[1;32m    704\u001b[0m \u001b[38;5;66;03m#                            (self._address, address_dict['network'], self.network.name))\u001b[39;00m\n\u001b[1;32m    705\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mpublic_hash \u001b[38;5;241m=\u001b[39m address_dict[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mpublic_key_hash_bytes\u001b[39m\u001b[38;5;124m'\u001b[39m]\n",
      "\u001b[0;31mTransactionError\u001b[0m: Network for output address tb1qlj64u6fqutr0xue85kl55fx0gt4m4urun25p7q is different from transaction network. bitcoin not in ['testnet', 'testnet4', 'signet']"
     ]
    }
   ],
   "source": [
    "#TEST Transaction\n",
    "other_wallet = wallet_create_or_open('OtherTestWallet', network='testnet')\n",
    "\n",
    "other_key = other_wallet.get_key()\n",
    "\n",
    "# Address to send funds (another testnet address you control)\n",
    "receiver_address = other_key.address\n",
    "\n",
    "# Send amount (e.g., 0.0001 BTC)\n",
    "tx = wallet.send_to(receiver_address, amount='0.0001 BTC')\n",
    "\n",
    "#Send testnet bitcoin back to address tb1qlj64u6fqutr0xue85kl55fx0gt4m4urun25p7q\n",
    "\n",
    "print(\"Transaction sent! TXID:\", tx.txid)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a0d0106e",
   "metadata": {},
   "outputs": [],
   "source": [
    "from bitcoinlib.transactions import Transaction\n",
    "\n",
    "#CHECK Transaction\n",
    "# Replace with your real TXID\n",
    "txid = \"your_transaction_id_here\"\n",
    "tx_info = Transaction.get(txid, network='testnet')\n",
    "\n",
    "print(\"Transaction details:\")\n",
    "print(tx_info.as_dict())"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
